### ISSUE TITLE
Add alternative to ValidationMethod

### ISSUE THREAD

**Issue by manuel-hegner** [dropwizard/dropwizard#2140]
Hey,
I would like to propose an alternative to the somewhat clunky ValidationMethod included in dropwizard. If this discussion leads to something I would also be willing to implement this new feature.

My proposal is to enable more detailed annotations by allowing users to use validationMethods that look like this:
```java
    @NewValidationMethod
    public void validateSomething(ConstraintValidatorContext context) {}
```
This would allow users a better control over the validation (e.g. better messages and multiple violations in one method) and would no longer interfere with Jackson (requiring no @JsonIgnore).

To make this work every Class that has this kind of validation methods would need to be annotated with a new @HasAnnotationMethods annotation, which would be a real hibernate constraint and would find every new-type validation method and execute them.

A caveat that I see is performance (on the other hand hibernate validation performance is not great as it is). To not use slow reflection access, code generation would be the better alternative. Does dropwizard already use a library for that? I would not like adding new dependencies just for this feature.

What do you think?

**Comment by nickbabcock** [comment]
Based on your description, I'm not in favor of removing `ValidationMethod` as the proposed method doesn't seem to make it irrelevant.

However, I approve of adding the proposed method to complement the current validation method. If this can fit within the current structure of the dropwizard then we can accept it as is. Else if third party dependencies need to be pulled in or if there is a significant performance penalty for those who don't opt-in to the new method, then I'd recommend making another module for encapsulation.

**Comment by manuel-hegner** [comment]
Okay, I started looking at how to implement this. I would need two dependencies to implement this cleanly in dropwizard-validation. The first one is slf4j, which I think would not be a problem. The other one is Javassist to implement the code generation. This is already a transitive dependency of dropwizard-jersey (through jersey and hk2-locator). I am not sure if it would be okay to add this dependency explicitely to dropwizard-validation. What are your thoughts?

**Comment by isaki-x** [comment]
There are alternative solutions to the code generation problem. For example [ReflectASM](https://github.com/EsotericSoftware/reflectasm) has proven to be pretty performant (it is used by Kryo, which makes sense since the same authors maintain Kryo and ReflectASM).

**Comment by manuel-hegner** [comment]
Yeah, Reflectasm is definitely the better choice for this kind of problem. I just thought that with javassist all of this could be done within java-validation. I don't really see anyone using this if it is in a new project.

**Comment by evnm** [comment]
As @nickbabcock suggested, a good way to move forward would be to implement this functionality as a module outside of dropwizard-validation. Treating such a module as a proof-of-concept, we then decide whether to incorporate the new validation annotation as a mainline contribution.

**Comment by arteam** [comment]
Closed via #2150
